src/lang-go.ts

```diff
@@ -727,63 +660,149 @@ export async function activateUsingWebSockets(ctx: sourcegraph.ExtensionContext)
     async function afterActivate(): Promise<void> {
-        const address = sourcegraph.configuration.get<Settings>().get('go.serverUrl')
-        if (address) {
-            ctx.subscriptions.add(registerFeedbackButton({ languageID: 'go', sourcegraph, isPrecise: true }))
-
-            await activateUsingWebSockets(ctx)
-        } else {
-            ctx.subscriptions.add(
-                registerFeedbackButton({
-                    languageID: 'go',
-                    sourcegraph,
-                    isPrecise: false,
-                })
-            )
+        const lsif = initLSIF()
+        const lsp = await initLSP(ctx)
+        const basicCodeIntel = initBasicCodeIntel()
 
-            activateBasicCodeIntel({
-                sourcegraph,
-                languageID: 'go',
-                fileExts: ['go'],
-                filterDefinitions: ({ repo, filePath, pos, fileContent, results }) => {
-                    const currentFileImportedPaths = fileContent
-                        .split('\n')
-                        .map(line => {
-                            // Matches the import at index 3
-                            const match = /^(import |\t)(\w+ |\. )?"(.*)"$/.exec(line)
-                            return match ? match[3] : undefined
-                        })
-                        .filter((x): x is string => Boolean(x))
-
-                    const currentFileImportPath = repo + '/' + path.dirname(filePath)
-
-                    const filteredResults = results.filter(result => {
-                        const resultImportPath = result.repo + '/' + path.dirname(result.file)
-                        return (
-                            currentFileImportedPaths.some(i => resultImportPath.includes(i)) ||
-                            resultImportPath === currentFileImportPath
-                        )
-                    })
-
-                    return filteredResults.length === 0 ? results : filteredResults
-                },
-                commentStyle: {
-                    lineRegex: /\/\/\s?/,
-                },
-            })(ctx)
-        }
+        ctx.subscriptions.add(
+            sourcegraph.languages.registerHoverProvider(goFiles, {
+                provideHover: asyncFirst([lsif.hover, lsp.hover, wrapMaybe(basicCodeIntel.hover)], null),
+            })
+        )
+        ctx.subscriptions.add(
+            sourcegraph.languages.registerDefinitionProvider(goFiles, {
+                provideDefinition: asyncFirst(
+                    [lsif.definition, lsp.definition, wrapMaybe(basicCodeIntel.definition)],
+                    null
+                ),
+            })
+        )
+        ctx.subscriptions.add(
+            sourcegraph.languages.registerReferenceProvider(goFiles, {
+                provideReferences: asyncFirst(
+                    [lsif.references, lsp.references, wrapMaybe(basicCodeIntel.references)],
+                    null
+                ),
+            })
+        )
```

Comment:
Here's the core of this change. The idea is to try LSIF, then language server, then basic-code-intel in that order. If one of them succeeds, stop and use its result.

`asyncFirst` tries calling each function in the array in turn, stopping when it got a non-`undefined` result. `null` is the default value (in case none of the functions returned non-`undefined`, which is actually impossible because the `basicCodeIntel` functions always return a result).